### PR TITLE
Fix bugs I found

### PR DESCRIPTION
**Pull request by lldelisle**
I added a lot of comments
I fixed 3 bugs:
1. when no circle where found because the minimum radius was too large, it used to recalculate circles with minimum_radius of 10. This is fixed by cf11b9e77fdc8b4ba87da5b9078ca1c72abbe799
2. if you put a high value for closeness tolerance, it was considering as adjacent circle identical circles. This is fixed by 5f1b2551a58b937e04715a96e56c9cc6a0084ef3
3. it was considered adjacent 2 circles with distance below tolerance even if you needed to go over a part which is not in mask:
image
I forced adjacent circles to have the full line between centers in mask (maybe this point needs to be discussed) in 10c728e446158a7b36bd42678bafffb87f3499ed
4. When 2 circles where found no spine was computed this is fixed by 75cb07e417da82d98c551272d320d814e8a0dc05

Points to discuss:
- I replaced the test: `c1.distance(c0) > r1 + r0 - this.closenessTolerance` by `c1.distance(c0) > 0`. Is it ok?
- The point 3 above especially if we work with selection instead of mask...
- Should we keep the log with compatible adjacent circles and spine coordinates?

Macro to test:
```ijm
newImage("Only2Circles", "8-bit black", 512, 512, 1);
//setTool("polygon");
makePolygon(80,195,256,53,366,183,111,286);
setForegroundColor(255, 255, 255);
run("Fill", "slice");
run("Max Inscribed Circles", "minimum=100 get minimum_0=0.50 closeness=5");
newImage("NoSpineBecauseOverlapBlack", "8-bit black", 512, 512, 1);
//setTool("polygon");
makePolygon(171,48,438,56,429,273,227,241,250,312,386,314,364,488,207,382);
setForegroundColor(255, 255, 255);
run("Fill", "slice");
run("Max Inscribed Circles", "minimum=100 get minimum_0=0.50 closeness=500");
newImage("NoSpineBecauseSingleCircle", "8-bit black", 512, 512, 1);
//setTool("polygon");
makePolygon(171,48,438,56,429,273,227,241,250,312,386,314,364,488,207,382);
setForegroundColor(255, 255, 255);
run("Fill", "slice");
run("Max Inscribed Circles", "minimum=200 get minimum_0=0.50 closeness=500");
newImage("NoSpineBecauseNoCircle", "8-bit black", 512, 512, 1);
//setTool("polygon");
makePolygon(171,48,438,56,429,273,227,241,250,312,386,314,364,488,207,382);
setForegroundColor(255, 255, 255);
run("Fill", "slice");
run("Max Inscribed Circles", "minimum=500 get minimum_0=0.50 closeness=500");
newImage("ShouldWork", "8-bit black", 512, 512, 1);
//setTool("polygon");
makePolygon(171,48,438,56,429,273,227,241,250,312,386,314,364,488,207,382);
setForegroundColor(255, 255, 255);
run("Fill", "slice");
run("Max Inscribed Circles", "minimum=50 get minimum_0=0.50 closeness=500");
newImage("ShouldWorkEvenBetter", "8-bit black", 512, 512, 1);
//setTool("polygon");
makePolygon(171,48,438,56,429,273,227,241,250,312,386,314,364,488,207,382);
setForegroundColor(255, 255, 255);
run("Fill", "slice");
run("Max Inscribed Circles", "minimum=30 get minimum_0=-0.5 closeness=500");

```
image

### CODE REVIEW

**Comment by lacan** [comment]
Hi, 
Awesome thanks for taking the time to do all this.

A couple of comments were added to the commits you mentioned. Happy to discuss it further. 

I would kindly offer that you add yourself to the contributors in the pom.xml file as well if that's OK with you.
```
        <contributor>
            <name>Lucille Delisle</name>
            <url>https://people.epfl.ch/lucille.delisle</url>
            <roles>
                <role>debugger</role>
                <role>reviewer</role>
            </roles>
            <properties>
                <id>lldelisle</id>
            </properties>
        </contributor>
```

**Comment by lldelisle** [comment]
Can you confirm:
- I replaced the test: c1.distance(c0) > r1 + r0 - this.closenessTolerance by c1.distance(c0) > 0. This is ok, right?
- Should we keep the log with compatible adjacent circles and spine coordinates?
After running the test macro the output is:
```
 
 spinepoints [Point2D [x = 309.4556884765625, y = 115.08023834228516]]

 spinepoints [Point2D [x = 95.55675506591797, y = 244.89974975585938]]
Error: No Adjacent Circles found. Consider decreasing 'Minimum Circle Diameter' or increasing 'Closeness Tolerance'
No spine found
A single circle was found. Spine cannot be computed, consider decreasing 'Minimum Circle Diameter'.
No circles found, consider decreasing 'Minimum Circle Diameter'.

 spinepoints [Point2D [x = 432.433837890625, y = 213.61607360839844]]

 spinepoints [Point2D [x = 171.58575439453125, y = 59.385257720947266]]
[Roi[Oval, x=387, y=226, width=43, height=43], Roi[Oval, x=400, y=55, width=37, height=37]]

 spinepoints [Point2D [x = 408.5, y = 247.50000122398885], Point2D [x = 428.0289611816406, y = 273.5139465332031]]
[Roi[Oval, x=181, y=120, width=48, height=48]]
[Roi[Oval, x=201, y=304, width=58, height=58], Roi[Oval, x=188, y=186, width=56, height=56], Roi[Oval, x=196, y=262, width=43, height=43], Roi[Oval, x=387, y=226, width=43, height=43], Roi[Oval, x=400, y=55, width=37, height=37], Roi[Oval, x=215, y=361, width=34, height=34]]
[Roi[Oval, x=249, y=312, width=128, height=128], Roi[Oval, x=328, y=432, width=40, height=40], Roi[Oval, x=215, y=361, width=34, height=34]]
[Roi[Oval, x=328, y=432, width=40, height=40]]

 spinepoints [Point2D [x = 204.99999953171323, y = 144.0000009302396], Point2D [x = 230.00000093108508, y = 333.00000058610345], Point2D [x = 312.999998202124, y = 375.9999977534402], Point2D [x = 348.00000194390117, y = 452.0000019439014], Point2D [x = 362.7868957519531, y = 487.361083984375]]

```

File: src/main/java/ch/epfl/biop/CirclesBasedSpine.java [hunk 1]
```diff
@@ -103,34 +116,47 @@ float[] toFloatArray(double[] arr) {
...
                 this.ov.add(line);
+                // add the extremity point to the spinePoints
                 spinePoints.add(new Point2D((double)line.getFloatPolygon().xpoints[2], (double)line.getFloatPolygon().ypoints[2]));
             }
         }
-
+        // set this.ov as overlay
         this.imp.setOverlay(this.ov);
+        // Print in log the spine points (from circleB excluded)
         IJ.log("\n spinepoints " + spinePoints);
```
Comment (lacan) [inline src/main/java/ch/epfl/biop/CirclesBasedSpine.java]: We can remove the log for now, I will add a verbose mode later on

File: src/main/java/ch/epfl/biop/CirclesBasedSpine.java [hunk 2]
```diff
@@ -103,34 +116,47 @@ float[] toFloatArray(double[] arr) {
...
-            List theCircles = (List)this.getAdjacentCircles(circles, circleB).stream().filter((Roi c) -> {
+            // Find a circle adjacent to circleB which is not A
+            // Which is compatible with similarity
+            List<Roi> theCircles = (List<Roi>)this.getAdjacentCircles(circles, circleB, imp).stream().filter((Roi c) -> {
                 Point2D vectorB = this.getVector(finalCircleB, c);
                 return this.similarity(vectorA, vectorB) > this.minSimilarity;
             }).collect(Collectors.toList());
             if (!theCircles.isEmpty()) {
+                // Write to log all the compatible adjacent circles
                 IJ.log("" + theCircles);
```
Comment (lacan) [inline src/main/java/ch/epfl/biop/CirclesBasedSpine.java]: We can also remove these

**Comment by lacan** [comment]
> I replaced the test: c1.distance(c0) > r1 + r0 - this.closenessTolerance by c1.distance(c0) > 0. This is ok, right?

Yes I believe so!

**Comment by lldelisle** [comment]
Then for me it is perfect.

**Comment by lldelisle** [comment]
Much cleaner:
```
Error: No Adjacent Circles found. Consider decreasing 'Minimum Circle Diameter' or increasing 'Closeness Tolerance'
No spine found
A single circle was found. Spine cannot be computed, consider decreasing 'Minimum Circle Diameter'.
No circles found, consider decreasing 'Minimum Circle Diameter'.

```

**Comment by lacan** [comment]
I will make a release tomorrow! Thanks for the contributions.
Best,
Oli

**Comment by lldelisle** [comment]
Would you mind to ping me when the release is done? Do you plan to put it available on maven?

**Comment by lacan** [comment]
Hello @lldelisle !

We don't release to Maven Central but to Maven SciJava, as it it quite dependent on other bits, like Fiji

It took a bit, and a few more commits to make the POM SciJava Friendly, but we now have an artifact 1.1.2 that is released and under CI

[https://maven.scijava.org/#nexus-search;gav~ch.epfl.biop~MaxInscribedCircles~1.1.2~~](https://maven.scijava.org/#nexus-search;gav~ch.epfl.biop~MaxInscribedCircles~1.1.2~~)

```
    <repositories>
        <repository>
            <id>scijava.public</id>
            <url>https://maven.scijava.org/content/groups/public</url>
        </repository>
    </repositories>
```

**Comment by lldelisle** [comment]
Hi @lacan ,
The release is in maven.scijava but not in http://biop.epfl.ch/Fiji-Update/plugins/, is it on purpose?

**Pull request merged**